File: 23/part2.rb
Node = Struct.new(:val, :next)

head = nil
tail = nil

cups = '685974213'.split('').map { Node.new(_1.to_i, nil) }
head = cups.first
tail = cups.first
cups[1..-1].each do |c|
  tail.next = c
  tail = c
end

(10..1_000_000).each do |i|
  c = Node.new(i, nil)
  tail.next = c
  tail = c
end

pointers = Array.new(1_000_001)
z = head
loop do
  pointers[z.val] = z
  break if z.next.nil?
  z = z.next
end

10_000_000.times do |i|
  curr = head
  head = curr.next

  a = head
  head = a.next

  b = head
  head = b.next

  c = head
  head = c.next

  pickup = [a, b, c]
  pickup_vals = pickup.map { _1.val }

  dest = curr.val
  loop do
    dest -= 1
    dest = 1_000_000 if dest == 0
    break unless pickup_vals.include?(dest)
  end

  finder = pointers[dest]

  after_finder = finder.next
  finder.next = a
  c.next = after_finder

  tail = c if after_finder.nil?

  curr.next = nil
  tail.next = curr
  tail = curr
end

one_node = pointers[1]
a = one_node.next
b = a.next
puts a.val, b.val, (a.val)*(b.val)

File: 23/part2.swift
import Foundation

class Node {
    let val: Int
    var next: Node?

    init(val: Int, next: Node?) {
        self.val = val
        self.next = next
    }
}

var cups = [Node]()
for c in "685974213" {
    cups.append(Node(val: Int("\(c)")!, next: nil))
}

var head: Node = cups[0]
var tail: Node = cups[0]

for c in cups.dropFirst() {
    tail.next = c
    tail = c
}

for i in (10...1000000) {
    let c = Node(val: i, next: nil)
    tail.next = c
    tail = c
}

var pointers: [Node] = Array(repeating: head, count: 1000001)

var z = head
while true {
    pointers[z.val] = z
    if let n = z.next {
        z = n
    } else {
        break
    }
}

for _ in (1...10000000) {
    let curr = head
    head = curr.next!

    let a = head
    head = a.next!

    let b = head
    head = b.next!

    let c = head
    head = c.next!

    let pickup = [a, b, c]
    let pickup_vals = pickup.map { n in n.val }

    var dest = curr.val
    while true {
        dest -= 1
        if dest == 0 {
            dest = 1000000
        }
        if !pickup_vals.contains(dest) {
            break
        }
    }

    let finder = pointers[dest]

    let after_finder = finder.next
    finder.next = a
    c.next = after_finder

    if after_finder == nil {
        tail = c
    }

    curr.next = nil
    tail.next = curr
    tail = curr
}

let one_node = pointers[1]
let a = one_node.next!
let b = a.next!

print(a.val)
print(b.val)
print(a.val * b.val)
